### PR TITLE
 Added Support for Structured Bindings 

### LINKED ISSUE #1388: Adding Structured Binding Support
- Describe the feature in as much detail as possible.
Given the fact that the Json Library now implements `items()` function, utilising [Structured Bindings](https://en.cppreference.com/w/cpp/language/structured_binding) in this scenario would make the code much more simpler, easier to read and understand
Even though Structured Bindings were added in C++17, the components/engines which ensure it exists are present in C++11 and C++14 and adding support would not require any `#ifdef `

- Include sample usage where appropriate.
 A Sample usage of Structured Binding is

```
   // example for an object
   for (auto& [key,value] : j_object.items())
   {
      std::cout << "key: " << key << ", value: " << value << '\n';
   }
```

- Implementation
The reason I am writing it as an issue rather than as a pull request is because it makes one questionable change.
Most of the implementation is referenced from [TartanLlama's Blog
](https://blog.tartanllama.xyz/structured-bindings/)
In the Class `iteration_proxy_internal` we would have to add the function get for Structured Bindings Support.
Thanks for @theodelrieu 's recommendations which helped shorten the code

```
// Required for Structured Bindings support
template<std::size_t N, typename = typename std::enable_if<N >= 2>::type>
void get()const
{
   static_assert(false, "Has Max 2 Elements:- Key or Value");
}
template<std::size_t N, typename = typename  std::enable_if<N == 0>::type>
auto get() const -> decltype(key())
{
   return key();
}
template<std::size_t N, typename = typename std::enable_if<N == 1>::type>
auto get() const  -> decltype(value())
{
   return value();
}
```
This is not a major change and would cause no incompatibility.
Adding support for Structured Bindings however involves minor editing of the std namespace which may or may not be favoured by a few projects.

```
namespace std {
   template<std::size_t N>
   struct tuple_element<N, nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal> {
      using type = decltype(std::declval<nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal>().get<N>());
   };

   template<>
   struct tuple_size<nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal>
      : std::integral_constant<std::size_t, 2> {};
}
```
This is where we face an issue.
In both these methods, iteration_proxy_internal needs to be accessible.
However, as we know, iterator_proxy_internal is defined as a `private` method in the `iterator_proxy` class.
Hence in order to add support for Structured Bindings, we would have to either make it public, find a hack or use something similar to `friend`
This is also the reason I decided to submit it as an issue rather than make a pull request.

I apologise for any and all mistakes committed by me while writing this.
I am quite new to GitHub and this is one of my first feature requests on a project

EDIT: Thanks to @theodelrieu for making recommendations which helped shorten the code

### PR DESCRIPTION
**Pull request by pratikpc**
Closes #1388.

[Describe your pull request here. Please read the text below the line, and make sure you follow the checklist.]

```
   for (auto const& [key,value] : json.items())
   {
      std::cout << "key: " << key << ", value: " << value << '\n';
   }
```
For further details, read #1388 and https://blog.tartanllama.xyz/structured-bindings/

Structured Bindings help simplify writing code and using this library greatly especially given the way they are implemented in C++17 and the addition of the items function.
The method of implementation ensures that Structured Bindings implementation can also be added in projects using C++11 with only C++17 being able to reap the benefits.

* * *

## Pull request checklist

Read the [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md) for detailed information.

- [ ]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [ ]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [ ]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src).
- [ ]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header file `single_include/nlohmann/json.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

## Please don't

- The C++11 support varies between different **compilers** and versions. Please note the [list of supported compilers](https://github.com/nlohmann/json/blob/master/README.md#supported-compilers). Some compilers like GCC 4.7 (and earlier), Clang 3.3 (and earlier), or Microsoft Visual Studio 13.0 and earlier are known not to work due to missing or incomplete C++11 support. Please refrain from proposing changes that work around these compiler's limitations with `#ifdef`s or other means.
- Specifically, I am aware of compilation problems with **Microsoft Visual Studio** (there even is an [issue label](https://github.com/nlohmann/json/issues?utf8=✓&q=label%3A%22visual+studio%22+) for these kind of bugs). I understand that even in 2016, complete C++11 support isn't there yet. But please also understand that I do not want to drop features or uglify the code just to make Microsoft's sub-standard compiler happy. The past has shown that there are ways to express the functionality such that the code compiles with the most recent MSVC - unfortunately, this is not the main objective of the project.
- Please refrain from proposing changes that would **break [JSON](http://json.org) conformance**. If you propose a conformant extension of JSON to be supported by the library, please motivate this extension.
- Please do not open pull requests that address **multiple issues**.

### CODE REVIEW

File: include/nlohmann/detail/iterators/iteration_proxy.hpp [hunk 1]
```diff
@@ -99,11 +99,38 @@ template<typename IteratorType> class iteration_proxy
...
+        template<std::size_t N, typename = typename  std::enable_if<N == 0>::type>
+        auto get() const -> decltype(key())
+        {
+            return key();
+        }
+        template<std::size_t N, typename = typename std::enable_if<N == 1>::type>
+        auto get() const  -> decltype(value())
+        {
+            return value();
+        }
```
Comment (theodelrieu) [inline include/nlohmann/detail/iterators/iteration_proxy.hpp]: I think it's better to make those non-member, to avoid users calling them (not even talking about the possible confusion with `json::get`)

Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: That might not work.
Because the basic idea with Structured Bindings is that they return key and value
Key at Index 0
And Value at Index 1
At least when used with items.
Making them public/anything else would render them incompatible with Structured Bindings support in json.items

The issue was with iteration_proxy_internal being declared as private however, I have made it public so it should be compatible now

File: include/nlohmann/detail/iterators/iteration_proxy.hpp [hunk 2]
```diff
@@ -99,11 +99,38 @@ template<typename IteratorType> class iteration_proxy
...
     };
 
     /// the container to iterate
     typename IteratorType::reference container;
 
+    // In order to Support Structured Bindings
+    // We would have to make iteration_proxy_internal
+    // Accessible to tuple_size and tuple_element in std namespace
+    // However, as iteration_proxy_internal is private,
+    // We will have to use a Hack to add Structured Bindings Support
```
[the remark this answers is not in the document]
Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: I made the necessary changes

[the remark this answers is not in the document]
Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: I do not have much of a clue how LGTM or AppVeyor work so do they restart with the new commit or would I have to create a new one??

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 1 above)
Comment (theodelrieu) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: About non-members, the prototype is the same, but it takes an argument, where the member-function takes none. It should work the same either way.

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (theodelrieu) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: You do not have to create a new one IIRC

**Comment by pratikpc** [comment]
```
   for (auto const& [key,value] : json.items())
   {
      std::cout << "key: " << key << ", value: " << value << '\n';
   }
```

Is probably same as

```
   for (auto const& elem : json.items())
   {
      auto const key = elem.get<0>();
      auto const value = elem.get<1>();
      std::cout << "key: " << key << ", value: " << value << '\n';
   }
```

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 1 above)
Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: Not really sure.
Because as far as I know, the way Structured Bindings work is that they consider everything is a Tuple/Tuple Element.

For example Structured Binding related to the code

```
   for (auto const& [key,value] : json.items())
   {
      std::cout << "key: " << key << ", value: " << value << '\n';
   }

```
Is probably same as
```
   for (auto const& elem : json.items())
   {
      auto const key = elem.get<0>();
      auto const value = elem.get<1>();
      std::cout << "key: " << key << ", value: " << value << '\n';
   }
```

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: @theodelrieu thanks!
That's actually a nice feature

**Comment by theodelrieu** [comment]
In generic context, one would always use `std::get<0>(...)`, especially to avoid writing `j.template get<0>()`

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 1 above)
Comment (theodelrieu) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: From [cppreference](https://en.cppreference.com/w/cpp/language/structured_binding):

* `e.get<i>()`, if lookup for the identifier get in the scope of `E` by class member access lookup finds at least one declaration that is a function template whose first template parameter is a non-type parameter.

* Otherwise, `get<i>(e)`, where `get` is looked up by argument-dependent lookup only, ignoring non-ADL lookup.

Your assumption is correct when the first bullet point is matched, but when the second one is matched, code is equivalent to:

```cpp
for (auto const& elem : json.items())
{
    auto const key = std::get<0>(elem);
    auto const value = std::get<1>(elem);
    std::cout << "key: " << key << ", value: " << value << '\n';
}
```

Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: > 
> 
> In generic context, one would always use `std::get<0>(...)`, especially to avoid writing `j.template get<0>()`

I apologise.
It seems I can push it out to std namespace
The code would look similar to

```
namespace std {

   // Required for Structured Bindings support
   template<std::size_t N, typename = typename std::enable_if<N >= 2>::type>
   void get(nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal&)
   {
      static_assert(N >= 2, "Has Max 2 Elements:- Key or Value");
   }
   template<std::size_t N, typename = typename  std::enable_if<N == 0>::type>
   auto get(nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal const& x)  -> decltype(x.key())
   {
      return x.key();
   }
   template<std::size_t N, typename = typename std::enable_if<N == 1>::type>
   auto get(nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal const& x)   -> decltype(x.value())
   {
      return x.value();
   }

   template<std::size_t N>
   struct tuple_element<N, nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal> {
      using type = decltype(get<N>(std::declval<nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal const&>()));
   };

   template<>
   struct tuple_size<nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal>
      : std::integral_constant<std::size_t, 2> {};
}
```

We still however would require iteration_proxy_internal to be public and json::iterator to be a parameter which doesn't seem to be so bad.

Comment (theodelrieu) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: What is the compiler error? Could you show the full snippet?

Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: error C2672: 'get': no matching overloaded function found
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\utility(482): note: see reference to class template instantiation 'std::tuple_element<0,_Tuple>' being compiled
          with
          [
              _Tuple=nlohmann::detail::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer>>>::iteration_proxy_internal
          ]
JSONTest.cpp(10): note: see reference to class template instantiation 'std::tuple_element<0,const nlohmann::detail::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer>>>::iteration_proxy_internal>' being compiled
include\nlohmann\json.hpp(20236): error C2784: 'tuple_element<_Idx,std::pair<_Ty1,_Ty2>>::type &std::get(std::pair<_Ty1,_Ty2> &) noexcept': could not deduce template argument for 'std::pair<_Ty1,_Ty2> &' from 'const nlohmann::detail::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer>>>::iteration_proxy_internal'
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\utility(600): note: see declaration of 'std::get'

The error occurred in my test project

Comment (theodelrieu) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: Indeed, I forgot about that one, I guess you have to forward declare json::iterator then...

Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: Nope the new commit fails.
get function probably can't properly lie within the std namespace
It fails with an error that it was unable to find get when a Structured Binding is used

Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: I think we should just stick to the old commit that placed the get function in iterator_proxy

Comment (theodelrieu) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: I will take a look tomorrow, I think we're close

Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: Auto Deduction of Return Types also fails on a few older compilers so decided to copy paste return types there from the original method.
The issue with this is that if key/value's types are changed, then get's types need to be changed manually
https://travis-ci.org/nlohmann/json/jobs/466093802

**Comment by theodelrieu** [comment]
I just pushed a commit on my [fork](https://github.com/theodelrieu/json/tree/feature/structured_bindings) which is still a wip, but the basic use-case works (with non-member `get` functions).

I had to move out `iteration_proxy_internal` outside of `iteration_proxy` to be able to partially specialize `tuple_size` and `tuple_element` (otherwise, only `nlohmann::json` would support structured bindings, not different instances of `nlohmann::basic_json<...>`).

There might be some refactoring to be done with `iteration_proxy/_internal`, I'd like to know what @nlohmann thinks about it, I do not know that part of the code very well.

Also, I have not tested edge-cases yet...

@pratikpc What do you think about incorporating my changes in your PR? We can work out the remaining details if you want :)

**Comment by pratikpc** [comment]
> @pratikpc What do you think about incorporating my changes in your PR? We can work out the remaining details if you want :)

That's a superb change.
Making it public and independent of iteration_proxy kills of all the problems we were facing.
I am all for it.

In the commit we can possibly push the get function to the std namespace as well

```
namespace std
{
   // structured bindings support
   template <std::size_t N, typename IteratorType, enable_if_t<N == 0, int> = 0>
   auto get(const nlohmann::detail::iteration_proxy_internal<IteratorType>& i) -> decltype(i.key())
   {
      return i.key();
   }

   template <std::size_t N, typename IteratorType, enable_if_t<N == 1, int> = 0>
   auto get(const nlohmann::detail::iteration_proxy_internal<IteratorType>& i) -> decltype(i.value())
   {
      return i.value();
   }
}
```

Although I still believe we need to know @nlohmann 's views regarding that.

**Comment by pratikpc** [comment]
@theodelrieu the commit broke a Unit Test.
Specifically the test-regression

One way to correct it is to add a Constructor in Basic Json class which accepts iteration_proxy_internal as an argument
```
template<typename Iterator>
basic_json(detail::iteration_proxy_internal<Iterator> other) : basic_json{ other.value() }
{}
```

However, my belief is that this is a temporary fix

The error I faced was
`[2018-12-13 07:41:19] [build] /usr/include/c++/7/bits/stl_algo.h:5337:16: error: no match for ‘operator=’ (operand types are ‘std::insert_iterator<nlohmann::basic_json<> >’ and ‘nlohmann::detail::iteration_proxy_internal<nlohmann::detail::iter_impl<nlohmann::basic_json<> > >’)
`
Also faced [here at LGTM](https://lgtm.com/projects/g/nlohmann/json/logs/rev/pr-de6145de23d2ca77b3e54238002cba61a9b56205/lang:cpp/stage:Build%20merge%20(f789b0d))

**Comment by theodelrieu** [comment]
Oops, my bad, I only tested `unit-items.cpp`......

>In the commit we can possibly push the get function to the std namespace as well

That's illegal unfortunately, the `get` methods are where they should be, i.e. in the same namespace as the class.

I will make the code work on my box and let you know when I push!

**Comment by pratikpc** [comment]
> That's illegal unfortunately, the `get` methods are where they should be, i.e. in the same namespace as the class.

Oh.
I thought it would be fine if they were kept in std namespace.
Because Structured Bindings still works on GCC and MSVC
> 
> I will make the code work on my box and let you know when I push!
The unit-tests now no longer fail after adding the Constructor.
But yes, finding a more permanent solution would be a good idea for us all.

I would add a commit which shifts get back into the namespace as before

**Comment by theodelrieu** [comment]
I just push-forced on my fork, the fix was in `conversions/to_json.hpp` as you can see.
I've added a test with a custom `basic_json` alias as well, if you have other tests in mind do not hesitate to add them :)

>I thought it would be fine if they were kept in std namespace.

You can see the rules from [cppreference](https://en.cppreference.com/w/cpp/language/extending_std)

**Comment by pratikpc** [comment]
@theodelrieu I can't find the changes.
Did you make them at https://github.com/theodelrieu/json/blob/feature/structured_bindings/include/nlohmann/detail/conversions/to_json.hpp

**Comment by theodelrieu** [comment]
It is still in namespace `detail`, so the current name does not annoy me that much. Need @nlohmann's input on this one.

I do think the PR is complete, could you squash your commits?

**Comment by pratikpc** [comment]
@theodelrieu how does one squash a commit exactly???

**Comment by theodelrieu** [comment]
First, do not forget to run `make amalgamate` and commit the result.

Once this is done, run `git rebase -i develop`. It will open a prompt.
Then, keep the first line as-is, and replace `pick` in all the others by `fixup`.
Save and quit, then `git push --force`.

**Comment by pratikpc** [comment]
@theodelrieu thanks!!!

**Comment by pratikpc** [comment]
@theodelrieu The following tests FAILED:
	 80 - test-unicode_all (Timeout)
On Appveyor
Does this mean we have to start the process again?

**Comment by nlohmann** [comment]
I’ll restart.

**Comment by nlohmann** [comment]
Appveyor seems to work now. :)

**Comment by nlohmann** [comment]
I am traveling right now, so I won’t be able to check before next week. But anything that brings us closer toward support of structured bindings is greatly appreciated!

File: include/nlohmann/detail/iterators/iteration_proxy.hpp [hunk 3]
```diff
@@ -110,16 +111,45 @@ template<typename IteratorType> class iteration_proxy
...
+}
+
+template <std::size_t N, typename IteratorType, typename std::enable_if<N == 1, int>::type = 0>
+auto get(const nlohmann::detail::iteration_proxy_internal<IteratorType>& i) -> decltype(i.value())
+{
+    return i.value();
+}
 }  // namespace detail
 }  // namespace nlohmann
+
```
Comment (nlohmann) [inline include/nlohmann/detail/iterators/iteration_proxy.hpp]: Could you please add a comment why you are adding code to the `std` namespace?

File: include/nlohmann/detail/conversions/to_json.hpp [hunk 4]
```diff
@@ -290,31 +290,31 @@ void to_json(BasicJsonType& j, typename BasicJsonType::object_t&& obj)
 template <
     typename BasicJsonType, typename T, std::size_t N,
     enable_if_t<not std::is_constructible<typename BasicJsonType::string_t,
-                const T (&)[N]>::value,
+                const T(&)[N]>::value,
```
Comment (nlohmann) [inline include/nlohmann/detail/conversions/to_json.hpp]: Can you make sure you ran `make pretty` - the indentation and whitespace looks odd in the whole PR. This is not a blocker, just a nice-to-have.

File: include/nlohmann/detail/iterators/iteration_proxy.hpp [hunk 5]
```diff
@@ -3,104 +3,105 @@
...
 #include <nlohmann/detail/value_t.hpp>
+#include <nlohmann/detail/meta/type_traits.hpp>
 
 namespace nlohmann
 {
 namespace detail
 {
-/// proxy class for the items() function
-template<typename IteratorType> class iteration_proxy
+template <typename IteratorType> class iteration_proxy_internal
```
Comment (nlohmann) [inline include/nlohmann/detail/iterators/iteration_proxy.hpp]: Discussion: Maybe rename this to `iteration_proxy_value`?

File: include/nlohmann/detail/conversions/to_json.hpp (hunk 4 above)
Comment (pratikpc) [reply include/nlohmann/detail/conversions/to_json.hpp]: Sorry.
I was previously using MSYS 32-Bit which didn't have astyle installed so amalgamate worked without issues.
Switching to MSYS 64 Bit solves this issue.

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 5 above)
Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: Yes.
That sounds like a nice enough name and does provide information regarding what `iteration_proxy_internal `exactly does.

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 3 above)
Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: Would something like the following be fine?
```
// The Addition to the STD Namespace is required to add
// Structured Bindings Support to the iteration_proxy_value class
// For further reference see https://blog.tartanllama.xyz/structured-bindings/
```
Or would a TartanLlama link be inappropriate and I should add the less helpful cppreference link
https://en.cppreference.com/w/cpp/language/structured_binding

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 5 above)
Comment (theodelrieu) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: I'm ok with `iteration_proxy_value`

Comment (nlohmann) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: I also prefer `iteration_proxy_value` over `iteration_proxy_kv` or `iteration_proxy_key_value`.

Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: @nlohmann I corrected that and updated the branch with another Squash of the commits.
Could you please check and see if they are proper?
One of my many doubts is [here](https://github.com/nlohmann/json/pull/1391#discussion_r242005973)

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 3 above)
Comment (nlohmann) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: The comment is fine. Maybe also add a link to this PR as reference.

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 5 above)
Comment (nlohmann) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: I just added an answer.

File: include/nlohmann/detail/conversions/to_json.hpp (hunk 4 above)
Comment (nlohmann) [reply include/nlohmann/detail/conversions/to_json.hpp]: The diff still looks weird, but that's OK for now.

File: include/nlohmann/detail/iterators/iteration_proxy.hpp (hunk 3 above)
Comment (nlohmann) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: @pratikpc If you add this, I would be ready to merge.

Comment (pratikpc) [reply include/nlohmann/detail/iterators/iteration_proxy.hpp]: @nlohmann I made the required modifications

**Comment by nlohmann** [comment]
Thanks so much!
